libp2p-websocket/src/desktop.rs

```diff
@@ -0,0 +1,253 @@
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use futures::{stream, Future, IntoFuture, Sink, Stream};
+use multiaddr::{AddrComponent, Multiaddr};
+use rw_stream_sink::RwStreamSink;
+use std::io::{Error as IoError, ErrorKind as IoErrorKind};
+use swarm::Transport;
+use websocket::client::builder::ClientBuilder;
+use websocket::message::OwnedMessage;
+use websocket::server::upgrade::async::IntoWs;
+use websocket::stream::async::Stream as AsyncStream;
+
+/// Represents the configuration for a websocket transport capability for libp2p. Must be put on
+/// top of another `Transport`.
+///
+/// This implementation of `Transport` accepts any address that ends with `/ws`, and will try to
+/// pass the underlying multiaddress to the underlying `Transport`.
+///
+/// > **Note**: The `/wss` protocol isn't supported.
+#[derive(Debug, Clone)]
+pub struct WsConfig<T> {
+	transport: T,
+}
+
+impl<T> WsConfig<T> {
+	/// Creates a new configuration object for websocket.
+	///
+	/// The websockets will run on top of the `Transport` you pass as parameter.
+	#[inline]
+	pub fn new(inner: T) -> WsConfig<T> {
+		WsConfig { transport: inner }
+	}
+}
+
+impl<T> Transport for WsConfig<T>
+where
+	T: Transport + 'static, // TODO: this 'static is pretty arbitrary and is necessary because of the websocket library
+	T::RawConn: Send, // TODO: this Send is pretty arbitrary and is necessary because of the websocket library
+{
+	type RawConn = Box<AsyncStream>;
+	type Listener = stream::Map<
+		T::Listener,
+		fn((<T as Transport>::ListenerUpgrade, Multiaddr)) -> (Self::ListenerUpgrade, Multiaddr),
+	>;
+	type ListenerUpgrade = Box<Future<Item = Self::RawConn, Error = IoError>>;
+	type Dial = Box<Future<Item = Self::RawConn, Error = IoError>>;
+
+	fn listen_on(
+		self,
+		original_addr: Multiaddr,
+	) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+		let mut inner_addr = original_addr.clone();
+		match inner_addr.pop() {
+			Some(AddrComponent::WS) => {}
+			_ => return Err((self, original_addr)),
+		};
+
+		let (inner_listen, new_addr) = match self.transport.listen_on(inner_addr) {
+			Ok((listen, inner_new_addr)) => {
+				// Need to suffix `/ws` to the listening address.
+				let new_addr = inner_new_addr
+					.encapsulate("/ws")
+					.expect("the /ws suffix is always valid");
```

Comment:
Can we get a source for this? Why is `ws` always valid? What if in the future we do validation to prevent `/udp/.../ws`?

Reply:
The `encapsulate` method parses the parameter, and adds it to the end. An error can only happen here if we fail to parse the string `"/ws"`. I agree that parsing a string every time is bad, so I will change it.